### ISSUE TITLE
Recursive links and focus

### ISSUE THREAD

**Issue by Ross-Angus** [w3c/aria-practices#117]
To take a simple example, imagine a navigation bar:

```
<nav aria-label="Main navigation" id="main-nav" itemscope="" itemtype="https://schema.org/SiteNavigationElement">
    <ul role="menubar">
        <li role="menuitem"><a href="/">Home</a></li>
        <li role="menuitem"><a href="/about-us/">About us</a></li>
    </ul>
</nav>
```

… my default method of indicating the current page to the user would be as follows:

```
<nav aria-label="Main navigation" id="main-nav" itemscope="" itemtype="https://schema.org/SiteNavigationElement">
    <ul role="menubar">
        <li role="menuitem"><em title="You are here">Home</em></li>
        <li role="menuitem"><a href="/about-us/">About us</a></li>
    </ul>
</nav>
```

(the emphasis tag is used so that if no CSS was present, users would still be able to see the current page)

The problem is that the emphasis tag has no focus, so if the navigation included a second level, keyboard navigation to that second level would be impossible.

Is this the solution?

```
<nav aria-label="Main navigation" id="main-nav" itemscope="" itemtype="https://schema.org/SiteNavigationElement">
    <ul role="menubar">
        <li role="menuitem"><button title="You are here">Home</button></li>
        <li role="menuitem"><a href="/about-us/">About us</a></li>
    </ul>
</nav>
```

Or is it that recursive links are no longer an issue?

**Comment by mcking65** [comment]
Hi @Ross-Angus,

Putting a button inside a menuitem will not work. You can't nest widget roles like that unless the outer role is a composite widget role and the inner role is allowed in the context of the outer role. Button is not a valid child of menuitem. For example, menu is a composite and menuitem is a required context of menu or group. 

We are working on the APG menubar pattern now, but it is not clear to me that you want to build a menubar. 

Did you want the nav element to contain only a single tab stop? If the entire set of nav elements is going to be a composite widget that has only one tab stop, then menubar is one option to consider. Grid and toolbar are two others.

The menu pattern that is a work in progress is here:
http://w3c.github.io/aria-practices/#menu

On the other hande, if you are not going to 
script the navigation as a single composite widget, then remove all the aria roles and just leave the nav element as links.

BTW, a few things to watch: ...

1. The "nav" element creates a navigation landmark region. So, please do not include the word "navigation" in the lable. Doing so causes screen readers to say "Main navigation navigation region". Just label it "Main". Or, if your site is named "Ross", you may want to name it "Ross, which would then result in "Ross navigation".
2. Do not put ARIA widget roles on elements that are not focusable. So, if you were building a menu, the menuitem roles should be on the <A> elements and not on the <li> elements.
3. In ARIA 1.1, there is a new property called aria-current that you can use to indicate to assistive tech which element represents the currently displayed entity. It is not yet supported by browsers or AT yet though.

**Comment by Ross-Angus** [comment]
Thank you for taking the time to explain this.

I think I've only partially understood your answer, I'm afraid. I don't know the difference between widget roles and composite widget roles or what you mean by a single tab stop.

The error with the use of the role attribute was due to me following the work of [the A11y project](http://a11yproject.com/patterns/) (which I appreciate is not connected with the W3C). [Specifically, this example](http://codepen.io/grayghostvisuals/pen/ltjDa).

I assume (from my example) that 

`<li role="menuitem"><a href="/about-us/">About us</a></li>`

... should become:

`<li><a href="/about-us/" role="menuitem">About us</a></li>`

... even though a sub-menu unordered list would be a sibling of the `role="menuitem"`, rather than a child?

I don't think I've understood your point about recursive links. Should I use them, or not? While I appreciate that <button/> is the incorrect tab to use here, I don't understand how to use an element which can take focus but _won't_ cause recursive links.

Thank you for the information regarding the aria-label attribute. This is really useful! I'll be sure to implement it going forward.

I may be missing something, but I couldn't find your example markup for a menubar. I've probably overlooked the link in the documentation.

**Comment by mcking65** [comment]
@Ross-Angus 

The A11yProject [Accessible Dropdown Menu](http://codepen.io/grayghostvisuals/pen/ltjDa) 
isn't actually accessible. I don't have time to enumerate all the issues.

Instead, please have a look at our
[navigation menu example](http://w3c.github.io/aria-practices/examples/menubar/menubar-1/menubar-1.html) 
and please read the 
[menu design pattern](http://w3c.github.io/aria-practices/#menu).

I don't really know what you mean by recursive links. HTML does not allow the `A` element to contain other `A` elements.

Note in the APG examples that the `menuitem` elements that have submenus do contain their submenu. But, in those cases, it is necessary to label the menuitem element so that the name is not calculated from all the children. Also note that the elements that get focus are always a `menuitem`, `menuitemradio`, or `menuitemcheckbox`.

I hope this helps. I will close this issue. However, if you still have questions after studying the design pattern and working with the example code, feel free to re-open this issue.

**Comment by Ross-Angus** [comment]
What I meant by "recursive links" is links which cause the current page to refresh. I understood this was bad for accessibility, presumably because it would cause speech browsers to read the page from the start again. But I can't find this advice in the published guidelines.

Unfortunately, where I struggle with the ARIA working group is where I could find examples of markup. For example I can't find where the Github links to http://w3c.github.io/aria-practices/examples/menubar/menubar-1/menubar-1.html

While I appreciate that you do not have time to address the issues with the A11yProject, their code is easy to find, copy and parse. I'd really appreciate some more visibility on all the work your group is doing.

**Comment by mcking65** [comment]
@Ross-Angus

Yes, reloading the current page unnecessarily is not a good idea because focus will be lost. If a nav menu item is for the current page,          one useful approach is to have activation set focus to an appropriate place on the current page. Note, though, if not using the menu design pattern, then there may be better options.

You wrote:

> Unfortunately, where I struggle with the ARIA working group is where I could find examples of markup. For example I can't find where the Github links to
> http://w3c.github.io/aria-practices/examples/menubar/menubar-1/menubar-1.html

This concerns me. What happens when you go to that example page? Are you saying the link does not work for you? 

That page includes a functional menubar example with all the source, html, js, and css. And, It links back to the menu design pattern section of the main guide document in two places.

You continue:
> While I appreciate that you do not have time to address the issues with the A11yProject, their code is easy to find, copy and parse. I'd really appreciate some more visibility on all the work your group is doing.

Like a11yProject, we have a github repo with all the code. On the gh-pages site, Each example page displays its code as well as allows you to test drive it. We also publish drafts to the w3.org site.

If you have suggestions for improving visibility, please share.

**Comment by Ross-Angus** [comment]
Thank you for your reply. That information on how to deal with recursive links is useful and something I can implement instead of my button work-around going forward.

The links within the demo are not the issue - I'm sure they're all fine. But as a developer, what I'm after is link pointing in to demo pages from outside. This is perhaps more of an Information Architecture issue. For example, I would expect the ARIA Working Group and their example code to be mentioned here:

https://www.w3.org/WAI/ARIA/

It would also be useful for a list of the currently available ARIA demos to be linked to from this page (or a suitable child page):

https://www.w3.org/WAI/train

I now see that the authoring practices guide is the hub document for these examples:

http://w3c.github.io/aria-practices/

... a document which is mentioned on the ARIA page, but not linked to. This seems to be an oversight.

The most important link on the menu bar example page is the one which points back to the WAI-ARIA Authoring Practices guide. But it is included in the text "design pattern for menubar". Perhaps it should be given more prominence?

It's a minor point but it would be nice if a library such as https://highlightjs.org/ was used for the code samples, the code might be easier for humans to parse.

Thank you again for taking the time to reply. My main concern now is that useful work is difficult to find.
